src/Raven.Server/Documents/Patch/DictionaryDispose.cs

```diff
@@ -0,0 +1,189 @@
+﻿using System;
+using System.Collections.Generic;
+using V8.Net;
+
+namespace Raven.Server.Documents.Patch
```

Comment:
I really dislike all of those. 
Why is that needed, and why override the methods in such an awkward manner?
Why isn't this the responsibility of the owning object, instead of the collection itself?

Reply:
With V8.Net we deal with JS values by means of InternalHandle struct to avoid GC participation for better performance. When we store InternalHandle we should 'clone' it (increment reference count) and after use we should dispose it (decrement reference count and on getting zero dispose of the native JS value).
The generic JsHandle is struct as well to keep this approach.

With DictionaryDispose* helpers we get the centralization of this functionality when storing those handles as keys or values. RavenEtlScriptRun has seven of such dictionaries. Without them it turns to a mess and a high risk of multiple JS value leakages, that are really hard to find.

Reply:
We already have the scope of operation (`SingleRun`), we can control everything that goes in and out.
That scope can keep track of all those values and close them in a consistent manner.

Reply:
Following the discussion with Pawel I have refactored DictionaryDispose : Dictionary to DictionaryCloning : IDictionary to avoid methods overloading with 'new' and made RavenEtlScriptRun disposable to dispose of those dictionaries and other class members in time (as well as added disposing of _currentRun : RavenEtlScriptRun in RavenEtlDocumentTransformer).